Nuclinfo Major Pair and Minor Pair overhaul

ALescoulie opened the pull request:
Fixes #3720 

Changes made in this Pull Request:
 - Add a class for Major Pair and Minor Pair distance in the new overhauled nucleic acids module.


PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

ALescoulie committed:
commit major pair and minor pair classes

ALescoulie committed:
add minor pair test

ALescoulie committed:
commit major pair dist

pep8speaks commented:
Hello @ALescoulie! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2023-10-10 06:31:35 UTC

ALescoulie committed:
update CHANGELOG

ALescoulie committed:
bring up to PEP8

ALescoulie committed:
Merge branch 'MDAnalysis:develop' into nuclinfo_dist_rebuild

IAlibay commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+    Bases are matched by their index in the lists given as arguments.
+
+    Parameters
+    __________
```
(here and everywhere else) I know these still render, but the line style is very much different from the numpy-like style guide we use. Can you instead replace this with a `------` line?

ALescoulie replied:
fixed in a recent commit

IAlibay commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+        sel2: List[mda.AtomGroup] = []
+        strand = zip(strand1, strand2)
+
+        for s in strand:
+            if s[0].resname[0] in [c_name, t_name, u_name]:
+                a1, a2 = o2_name, c2_name
+            elif s[0].resname[0] in [a_name, g_name]:
+                a1, a2 = c2_name, o2_name
+            else:
+                raise ValueError(f"{s} are not valid nucleic acids")
```
Unless I'm misreading, this code block is essentially being repeated in all the child classes. Could you just use a staticmethod here that generalises this code?

orbeckst replied:
yes, please 👍

orbeckst replied:
And then you can test it separately.

ALescoulie replied:
refactored it into the `NucPairDist` parent class like @IAlibay recommended

IAlibay commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+            else:
+                raise ValueError(f"{s} are not valid nucleic acids")
+
+            sel1.append(s[0].atoms.select_atoms(f'name {a1}'))
```
You can double check this on your end but testing things out you'd probably get a ~ 5x speedup if you did
`s[0].atoms[np.where(s[0].atoms.names == a1)]` instead

richardjgowers replied:
I'd be curious if this was still true after the string interning changes...

orbeckst replied:
@richardjgowers PLEASE write a benchmark case to be run for anything that you improve

ALescoulie replied:
Has this been bench-marked yet? If not I can set one up and adjust the selection code accordingly.

IAlibay commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+            elif s[0].resname[0] in [a_name, g_name]:
+                a1, a2 = c2_name, o2_name
+            else:
+                raise ValueError(f"{s} are not valid nucleic acids")
```
Also please cover these errors with tests

ALescoulie replied:
I test in the refactor of the selection into `NucPairDist`

IAlibay commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+    Attributes
+    __________
+    results: numpy.ndarray
+    first index is selection second index is time
```
This is missing indentation it seems?

ALescoulie replied:
fixed

IAlibay commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+                raise ValueError(f"{s} are not valid nucleic acids")
+
+            sel1.append(s[0].atoms.select_atoms(f'name {a1}'))
+            sel2.append(s[1].atoms.select_atoms(f'name {a2}'))
```
What happens here if one of the selections returns an empty atomgroup and the other doesn't? (for some reason or another, maybe the wrong name selection was passed?)

ALescoulie replied:
I added a check for that in the static method in `NucPairDist`

IAlibay commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -56,6 +64,7 @@
 from .distances import calc_bonds
 from .base import AnalysisBase, Results
 from MDAnalysis.core.groups import Residue
+from .dihedrals import Dihedral
```
Probably missing something obvious, where is this used?

ALescoulie replied:
Oh that didn't get cut when I moved a commit around. I was working on the rebuilding another the torsion calculator and had imported that, I'll remove that

ALescoulie replied:
removed in latest push

IAlibay reviewed (changes requested):
Sorry only had the time/energy for a quick review.

Couple of extra things I couldn't add comments for but worth mentioning since it seems like there's going to be further changes to this code:
1. lines 132, 139, and 142 - we already populate a times array under self.times, is this still necessary to populate? Can we just do a call to `np.array(self.times)`?
2. The Attributes for NucPairDist aren't indented properly, could you change them here?
3. Dicussion point (other @MDAnalysis/coredevs please weigh in here), the way the `results` dictionary is being used here is very different from how we do results everywhere else & imho makes it a bit hard to specifically extract the pairs of distances vs say the times entry. I realise it's a bit late since this already shipped in 2.2.0, but before we keep going, would it make sense to consider switching this to instead be `results.pair_distances` (maybe even making it a pairs x times 2D ndarray?

IAlibay commented on testsuite/MDAnalysisTests/analysis/test_nucleicacids.py:
```diff
@@ -23,7 +23,8 @@
 import MDAnalysis as mda
 import pytest
-from MDAnalysis.analysis.nucleicacids import WatsonCrickDist
+from MDAnalysis.analysis.nucleicacids import WatsonCrickDist,\
+    MajorPairDist, MinorPairDist
```
It doesn't get picked up by pep8speaks, but implicit continuation is best where possible

```suggestion
from MDAnalysis.analysis.nucleicacids import (WatsonCrickDist,
                                              MajorPairDist, MinorPairDist,)
```

ALescoulie replied:
fixed the issue in my latest push

IAlibay commented on testsuite/MDAnalysisTests/analysis/test_nucleicacids.py:
```diff
@@ -43,3 +44,27 @@ def test_wc_dist(u):
+    MI = MinorPairDist(strand1, strand2)
+    MI.run()
+
+    assert_allclose(MI.results[0][0], 15.06506, atol=1e-3)
```
another one for @MDAnalysis/coredevs to consider for our testing style going ahead - single float comparisons are ~ 20x faster (at least on my laptop) if done using `pytest.approx` instead of `assert_allclose`. It's microseconds (170 microsec vs 9 microsec), but it adds up quickly to seconds, how do we feel about making pytest.approx the assert method of choice in those cases?

orbeckst replied:
I raised #3743 for you.

ALescoulie replied:
I refactored my test to use pytest approx

IAlibay commented on package/CHANGELOG:
```diff
@@ -27,6 +27,8 @@ Enhancements
   * Additional logger.info output when per-frame analysis starts (PR #3710)
   * Timestep has been converted to a Cython Extension type
     (CZI Performance track, PR #3683)
+  * Add higher performance AnalysisBase derived Major and Minor pair distances to
```
We usually go newest first for these entries, could you move it to the top of the enhancements list?

IAlibay reviewed (dismissed):
Couple of extra things since I was playing with pytest.approx

richardjgowers reviewed (commented):
Thanks for doing this!  Refactoring old code is never pleasant.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+class MinorPairDist(NucPairDist):
+    r"""Minor-Pair basepair distance for selected residues over a trajectory.
+
+    Takes two lists of :class:`~MDAnalysis.core.groups.Residue` objects and
```
Can these be `ResidueGroup`, too? That would tie in even better with the OO approach.

ALescoulie replied:
I'm looking at the source and docs for `ResidueGroup`. I can see the utility in merging `NucPairDist` with it, since it would allow for easier access to the residue related data in the classes. I'm a little unclear though on how it would handle the two strands the user provides, can `GroupBase` automatically deal with that if I add a line with `super(ResidueGroup, strand1, strand2)`.

orbeckst replied:
I didn't think that far ahead: All I thought was that instead of `strand1: List[Residue]` use `strand1: ResidueGroup` (etc).

orbeckst replied:
However, we then should also use ResidueGroup for WatsonCrick and then we're back to API changes... so if we use ResidueGroup here then we also need to deprecate `List[Residue]` for `WatsonCrick`.

I think I like `ResidueGroup` better because `strand = u.select_atoms("nucleic and resid 1-17").residues` looks better than `strand = list(u.select_atoms("nucleic and resid 1-17").residues)`. I _think_ one can simply convert a `res_list: List[Residue]` to `residues: [ResidueGroup]` with
```python
residues = ResidueGroup(res_list)
```
so accepting both for a transition period wouldn't be difficult.

ALescoulie replied:
Ok, I'll make it clear in my code that either can be used for these processes.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+    c2_name: str (optional)
+        Name of Carbon 2 of nucleic acids
+        by default assigned to N3
+    g_name: str (optional)
+        Name of Guanine in topology
+        by default assigned to G
+    a_name: str (optional)
+        Name of Adenine in topology
+        by default assigned to G
+    u_name: str (optional)
+        Name of Uracil in topology
+        by default assigned to U
+    t_name: str (optional)
+        Name of Thymine in topology
+        by default assigned to T
+    c_name: str (optional)
+        Name of Cytosine in topology
+        by default assigned to C
```
Would it be clearer to use `guanine="G"` instead of `g_name="G"` and same for everything else?

I find the meaning to be clearer when reading `guanine="G"`.

ALescoulie replied:
I updated it to use the full name, ex: `guanine_name=G`

orbeckst replied:
Now I see how we got to the API-breaking naming... my comment from a year ago when it would have been easy to change the API without a break. As I said elsewhere, `g_name` etc work.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+
+    Raises
+    ______
+    ValueError
+    if the residues given are not amino acids
+    ValueError
+    if the selections given are not the same length
```
indentation

ALescoulie replied:
fixed

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+
+    """
+
+    def __init__(self, strand1: List[Residue], strand2: List[Residue],
```
`Union(List[Residue], ResidueGroup)` ?

ALescoulie replied:
I'll add that, I looked at `ResidueGroup` and there is no reason `NucPairDist` can't accept it or `Residue`

ALescoulie replied:
I used a type alias to do it so that the code is less verbose, what do you think about that solution? It is a bit less clear in the docs, but simplifies the code.

orbeckst replied:
I don't understand what a type alias is and where you used it — sorry, apologies for my ignorance. Can you please explain and comment on the corresponding piece of code?

orbeckst replied:
At least from the docs it is not clear to me that the class would now also accept `strand1: ResidueGroup` — ... at least from your comment I surmise that's what _should_ be possible now.

ALescoulie replied:
@orbeckst A type alias is when you give a specific name to a composite type in my case `ResidueClass` is an alias for `Union[Residue, ResidueGroup]` so my code's type signatures are a bit less verbose

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+                 **kwargs) -> None:
+        sel1: List[mda.AtomGroup] = []
+        sel2: List[mda.AtomGroup] = []
+        strand = zip(strand1, strand2)
```
plural "strands" ?

orbeckst replied:
Or perhaps clearer, `double_strand`?

ALescoulie replied:
I set it to strands

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+        sel2: List[mda.AtomGroup] = []
+        strand = zip(strand1, strand2)
+
+        for s in strand:
```
rename `s` to `pair`  to make meaning clearer?

ALescoulie replied:
good idea, make the code more readable

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+        super(MinorPairDist, self).__init__(sel1, sel2, **kwargs)
+
+
+class MajorPairDist(NucPairDist):
```
most of my comments from `MinorPairDist` apply here, too

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+        First list of bases
+    strand2: List[Residue]
+        Second list of bases
+    o6_name: str (optional)
```
`oxygen6` ?

ALescoulie replied:
I updated all the names so that they are more clear

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+        Second list of bases
+    o6_name: str (optional)
+        Name of Oxygen 6 of nucleic acids
+        by default assigned to N1
```
Why is O6 assigned to N1? Do you mean that this is the distance? Needs to be written more clearly.

ALescoulie replied:
I copied the doc string and forgot to update the variable description, I fixed it in the last push

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+        by default assigned to N1
+    n4_name: str (optional)
+        Name of Nitrogen 4 of nucleic acids
+        by default assigned to N3
```
not clear why N4 is N3 by default ... is this what it's meant here??

ALescoulie replied:
its an error from copying the string

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,3 +218,155 @@ def __init__(self, strand1: List[Residue], strand2: List[Residue],
+    **kwargs: dict
+        arguments for :class:`~MDAnalysis.analysis.base.AnalysisBase`
+
+    Attributes
+    __________
+    results: numpy.ndarray
+    first index is selection second index is time
+    results.times: numpy.ndarray
+    times used in analysis
+
+    Raises
+    ______
+    ValueError
+    if the residues given are not amino acids
+    ValueError
+    if the selections given are not the same length
```
underlining and indentation

ALescoulie replied:
fixed

orbeckst reviewed (changes requested):
Thank you for the rewrite.

I have a bunch of readability comments (see inline).

I agree with @IAlibay that we should remove `results['times']`. I also agree that `results` needs restructuring, both for consistency but also for ease of accessing the data with e.g. mdacli. Let's raise a separate issue where others can also easily chime in. Normally we adhere to semantic versioning. However, I would consider this module still experimental and I am willing to break strict semver here and just change `results` without deprecations. The changes to results can go into a separate PR (preferrable) or this one (if a lot easier).

orbeckst commented:
Before going much further here we should discuss the format of `Results` — please see #3744.

IAlibay commented:
@ALescoulie It would be great to not let this work get forgotten, are you still looking to contribute this?

ALescoulie commented:
Yeah, I'll continue working on this. Sorry for abandoning it, I was busy with school and dealing with mental health issues. I'm getting the development repo set up on my desktop now.

orbeckst commented:
Hi @ALescoulie good to hear from you! Would be fantastic to get your work in!

Commit:
Merge branch 'develop' into nuclinfo_dist_rebuild

ALescoulie committed:
fix documentation formatting

ALescoulie committed:
refactor to use select_strand_atoms staticmethod

ALescoulie committed:
fix major and minor pair selections

ALescoulie committed:
refactor tests to use pytest approx

ALescoulie committed:
remove dihedral import

ALescoulie committed:
fix WCDist not passing all kwargs into selector

ALescoulie committed:
add test for empty AtomGroup error

ALescoulie commented:
I pushed a larger change to the PR. It was a bit hard picking up where I left off, but I was able to jump back into developing without too many issues. 

I refactored my code so that all of the Pair distance analysis classes use a `NucPairDist` static method `select_strand_atoms` which takes in the strands, selected atoms names, and the nucleic acid names. It then returns a tuple containing the the lists of atoms for each strand. It also has checks ensuring that the strands only contain Nucleic Acids and that the selections don't return empty `AtomGroups`. I also added tests for both errors.

### TODO
- [x] Reformat `NucPairDist` to use the newer results format
- [x] Fix misc docs issues
- [x] Update the change log

Commit:
update atom names in pair distance classes

Commit:
fix whitespace for pep8

Commit:
use typealias for union[residue, reiduegroup]

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -140,6 +148,41 @@ def __init__(self, selection1: List[mda.AtomGroup],
             self._s2 += selection2[i]
 
+    @staticmethod
+    def select_strand_atoms(
```
As a public facing method, needs some docs.

Include a `.. versionadded:: 2.7.0`.

orbeckst replied:
I like making it a separate method.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -164,19 +207,19 @@ class WatsonCrickDist(NucPairDist):
     strand2: List[Residue]
         Second list of bases
-    n1_name: str (optional)
+    nitrogen1_name: str (optional)
```
These parameter renamings are technically speaking an API break so we'd need to deprecate them, document deprecations, and test.

If you think the more explicit names are worth the effort then please do the above. Otherwise, I am also ok with reverting to the original kwargs. They are reasonably clear already and the docs say explicitly what they are.

ALescoulie replied:
I'll just revert to the older version to prevent API changes.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -210,23 +253,150 @@ class WatsonCrickDist(NucPairDist):
-                 n1_name: str = 'N1', n3_name: str = "N3",
-                 g_name: str = 'G', a_name: str = 'A', u_name: str = 'U',
-                 t_name: str = 'T', c_name: str = 'C',
+                 nitrogen1_name: str = 'N1', nitrogen3_name: str = "N3",
+                 guanine_name: str = 'G', adenine_name: str = 'A', uracil_name: str = 'U',
+                 tymine_name: str = 'T', cytosine_name: str = 'C',
```
See comment above — if you want to rename you'll also have to accept the old kwargs (possibly via `**kwargs`) for at least one minor release cycle.

ALescoulie replied:
I just reverted to the old kwargs

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -210,23 +253,150 @@ class WatsonCrickDist(NucPairDist):
-                a1, a2 = n1_name, n3_name
-            else:
-                raise ValueError(f"{s} are not valid nucleic acids")
+        selections: Tuple[List[mda.AtomGroup], List[mda.AtomGroup]] = self.select_strand_atoms(
```
Instead of `selections` I would call it `strand_atomgroups` or something similar (maybe just `strands`), so that it's clear that it contains actual objects and not just "selection strings".

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -210,23 +253,150 @@ class WatsonCrickDist(NucPairDist):
+        First list of bases
+    strand2: List[Residue]
+        Second list of bases
+    oxygen2_name: str (optional)
```
All these names must be consistent between `WatsonCrick` and the other classes, which means either deprecating the old kwargs in `WatsonCrick` _or_ changing the ones here to the current ones.

Ultimately, that's your choice, but it needs to be **consistent across the whole module**. You're defining the API for everything else so think about it how you want it to be. I am fine with either long or short names as long as the docs are clear (which they are) and everything is consistent and as uniform as possible so that it is a smooth user experience.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -210,23 +253,150 @@ class WatsonCrickDist(NucPairDist):
+    r"""Minor-Pair basepair distance for selected residues over a trajectory.
+
+    Takes two lists of :class:`~MDAnalysis.core.groups.Residue` objects and
+    calculates the Watson-Crick distance between them over the trajectory.
```
How is `MinorPairDist` different from `WatsonCrickDist`? The docs do not say. Please add details to the docs to describe the difference.

orbeckst replied:
Start by adding docs from orginal functions https://docs.mdanalysis.org/stable/documentation_pages/analysis/nuclinfo.html#MDAnalysis.analysis.nuclinfo.minor_pair

orbeckst replied:
Explicitly state what distances are calculated.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -210,23 +253,150 @@ class WatsonCrickDist(NucPairDist):
+class MajorPairDist(NucPairDist):
+    r"""Minor-Pair basepair distance for selected residues over a trajectory.
+
+    Takes two lists of :class:`~MDAnalysis.core.groups.Residue` objects and
+    calculates the Watson-Crick distance between them over the trajectory.
+    Bases are matched by their index in the lists given as arguments.
```
How is `MajorPairDist` different from `WatsonCrickDist`? The docs do not say. Please add details to the docs to describe the difference.

orbeckst replied:
Start by copying the original docs eg https://docs.mdanalysis.org/stable/documentation_pages/analysis/nuclinfo.html#MDAnalysis.analysis.nuclinfo.major_pair

Also, the docs here incorrectly say "Minor-Pair".

orbeckst replied:
Explicitly state what distances are calculated.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -210,23 +253,150 @@ class WatsonCrickDist(NucPairDist):
+        First list of bases
+    strand2: List[Residue]
+        Second list of bases
+    oxygen6_name: str (optional)
```
See comments above about consistent argument naming.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -210,23 +253,150 @@ class WatsonCrickDist(NucPairDist):
+    Attributes
+    ----------
+    results: numpy.ndarray
+    first index is selection second index is time
```
indent (I think)

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -210,23 +253,150 @@ class WatsonCrickDist(NucPairDist):
+    results: numpy.ndarray
+    first index is selection second index is time
+    results.times: numpy.ndarray
+    times used in analysis
```
indent (I think)

ALescoulie replied:
fixed in a prev commit

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -210,23 +253,150 @@ class WatsonCrickDist(NucPairDist):
+    ----------
+    results: numpy.ndarray
+    first index is selection second index is time
+    results.times: numpy.ndarray
```
should just be times (and not in results)

ALescoulie replied:
updated for all the classes

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -210,23 +253,150 @@ class WatsonCrickDist(NucPairDist):
+    ----------
+    results: numpy.ndarray
+        first index is selection second index is time
+    results.times: numpy.ndarray
```
should just be times (and not in results)

ALescoulie replied:
done

ALescoulie replied:
I uptated the docs for all the classes

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -210,23 +253,150 @@ class WatsonCrickDist(NucPairDist):
+
+    Attributes
+    ----------
+    results: numpy.ndarray
```
use results.distances or something like that in the Results dict

orbeckst replied:
Describe the output data structure.

ALescoulie replied:
updated the docs to expalin the arrays orginzation

ALescoulie replied:
I added a description to the docs for both

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -210,23 +253,150 @@ class WatsonCrickDist(NucPairDist):
+
+    Attributes
+    ----------
+    results: numpy.ndarray
```
Should be results.distances or something like that inside Results dict.

Describe the output data structure.

ALescoulie replied:
updated the docs to explain the arrays orginization

ALescoulie replied:
I updated the docs for all the classes describing how its a 2d numpy array with base pairs and frames as the indicies

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -210,23 +253,150 @@ class WatsonCrickDist(NucPairDist):
+        if the residues given are not amino acids
+    ValueError
+        if the selections given are not the same length
+
```
needs versionadded

ALescoulie replied:
added it on both major and minor pair

ALescoulie replied:
done

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -210,23 +253,150 @@ class WatsonCrickDist(NucPairDist):
+        if the residues given are not amino acids
+    ValueError
+        if the selections given are not the same length
+
```
needs versionadded

ALescoulie replied:
done

orbeckst reviewed (changes requested):
I like the compact code and moving the strand selection into a separate method. 

1. The primary issue is just that renaming kwargs is an API break and because the code has been in core for a while (and not marked as "Experimental" or anything like that) we'd need to properly deprecate any API changes. The alternative is using the current (old) names also for the newly added classes.
2. In terms of API coherency, I think it were better if actual MDA objects, namely `ResidueGroup`, instead of `list[Residue]` were accepted. This would require a deprecation of the current list of residues input although that would be worth it in my opinion, so that we have a seamless OO API.
3. There are also some documentation issues but these are easy to fix.

richardjgowers commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -140,6 +148,41 @@ def __init__(self, selection1: List[mda.AtomGroup],
+
+
+        for s in strand:
+            if s[0].resname[0] in pyrimidines:
+                a1, a2 = a2_name, a1_name
+            elif s[0].resname[0] in purines:
+                a1, a2 = a1_name, a2_name
```
This could probably get simplified to:
```python
if not s[0].resname[0] in pyrimidines_and_purines:
    raise ValueError
```
I'm not sure what assigning to a1 and a2 does here, so you could remove that step.

ALescoulie replied:
This is based on how it was done in nucinfo, I'm basing my code on that originonal material. As far as I know its so that the selections are specified to one side of the pair.

richardjgowers commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -140,6 +148,41 @@ def __init__(self, selection1: List[mda.AtomGroup],
+            ag1 = s[0].atoms.select_atoms(f'name {a1}')
+            ag2 = s[1].atoms.select_atoms(f'name {a2}')
+
+            if not all([len(ag) > 0 for ag in [ag1, ag2]]):
```
the `all` function doesn't require a list, so you could reduce this to `all(len(ag) for ag in [ag1, ag2])`

ALescoulie replied:
oh cool, I never knew that!

richardjgowers commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -140,6 +148,41 @@ def __init__(self, selection1: List[mda.AtomGroup],
+        
+        sel1: List[mda.AtomGroup] = []
+        sel2: List[mda.AtomGroup] = []
+        strand: Tuple[Residue, Residue] = zip(strand1, strand2)
```
the type annotations are a great addition.  I think this one (from `zip`) is technically a generator/iterator.  I think instead you want to annotate `s: Tuple[Residue, Residue]`

ALescoulie replied:
I just moved zip into the for look line.

Commit:
revert kwargs names to preserve old api

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -92,8 +106,6 @@ class NucPairDist(AnalysisBase):
 
     Attributes
     ----------
-    times: numpy.ndarray
-        Simulation times for analysis.
```
This should remain because it's created by AnalysisBase, isn't it?

ALescoulie replied:
fixed in the recent versions

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -183,8 +229,6 @@ class WatsonCrickDist(NucPairDist):
 
     Attributes
     ----------
-    times: numpy.ndarray
-        Simulation times for analysis.
```
Needs to stay, I think.

ALescoulie replied:
fixed in recent versions

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,24 +253,147 @@ class WatsonCrickDist(NucPairDist):
+        arguments for :class:`~MDAnalysis.analysis.base.AnalysisBase`
+
+    Attributes
+    ----------
```
add docs for `times` (thanks to AnalysisBase)

ALescoulie replied:
done

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -209,24 +253,147 @@ class WatsonCrickDist(NucPairDist):
+        arguments for :class:`~MDAnalysis.analysis.base.AnalysisBase`
+
+    Attributes
+    ----------
```
add docs for `times` thanks to AnalysisBase

ALescoulie replied:
done

ALescoulie commented:
> I like the compact code and moving the strand selection into a separate method.
> 
> 1. The primary issue is just that renaming kwargs is an API break and because the code has been in core for a while (and not marked as "Experimental" or anything like that) we'd need to properly deprecate any API changes. The alternative is using the current (old) names also for the newly added classes.
> 2. In terms of API coherency, I think it were better if actual MDA objects, namely `ResidueGroup`, instead of `list[Residue]` were accepted. This would require a deprecation of the current list of residues input although that would be worth it in my opinion, so that we have a seamless OO API.
> 3. There are also some documentation issues but these are easy to fix.

@orbeckst I think I have a good way to switch to using `ResidueGroup` from `List[Residue]` without API breaks. I'll make `NucPairDist` and`WCDist` accept both, but raise a `DepreciatedWarning` if you give it `List[Residue]` then just convert it internally. Than raise an issue tagged with 3,0 to remove support for residue lists when we do API changes for the major version update. What do you think of that solution?

Commit:
update NucPairDist.select_strand_atoms add docs

Commit:
refactor results dict and clarify result structure
in docs

Commit:
include versionadded in major and minor pair dists

orbeckst commented:
good idea!

ALescoulie commented:
@orbeckst I'll push that change today, was waiting for feedback on the idea before implementing it

Commit:
refactor WatsonCrickDist to raise
DeprecationWarning if List[Residue] given

Commit:
make error messages consistent

Commit:
update CHANGELOG

ALescoulie commented:
I think this one is pretty close to done, There might still be some misc typos in the docs, but aside from that I refactored the code based on the suggestions of @orbeckst to use `ResidueGroup` and pulled selections into their own static methods. I also have gotten full code coverage, and updated CHANGELLOG.

IAlibay commented:
@ALescoulie - could you please confirm that you are ok releasing this code under a LGPLv2.1+ license and that it adheres to the [developer certificate of origin](https://developercertificate.org/)?

Aside from this I will dismiss my review block and leave this to @orbeckst to review & merge.

ALescoulie commented:
> @ALescoulie - could you please confirm that you are ok releasing this code under a LGPLv2.1+ license and that it adheres to the [developer certificate of origin](https://developercertificate.org/)?
> 
> 
> 
> Aside from this I will dismiss my review block and leave this to @orbeckst to review & merge.

I am happy to release this under the LGPL license. I wrote all of this on my own based on the older nucinfo code, and the MDAnalysis framework. I hope that it helps people with their research.

orbeckst committed:
Merge branch 'develop' into nuclinfo_dist_rebuild

orbeckst reviewed (changes requested):
A test fails with issues importing TypeAlias https://github.com/MDAnalysis/mdanalysis/actions/runs/6059699690/job/16443750172?pr=3735 — please fix first.

You can ignore test failures related to #4209 in `test_creating_multiple_universe_without_offset`.

Commit:
fix type alias import issue

ALescoulie commented:
I just removed the type alias import, apparently that wasn't in Python 3.9, @orbeckst is there anything else to fix on this PR

ALescoulie committed:
Merge branch 'develop' into nuclinfo_dist_rebuild

orbeckst commented on package/CHANGELOG:
```diff
@@ -13,7 +13,9 @@ The rules for this file:
 ------------------------------------------------------------------------------
-??/??/?? IAlibay, ianmkenney, PicoCentauri, ALescoulie
+
+??/??/?? IAlibay, ianmkenney, PicoCentauri, pgbarletta, p-j-smith, richardjgowers, ALescoulie
```
line break
```suggestion
??/??/?? IAlibay, ianmkenney, PicoCentauri, pgbarletta, p-j-smith, richardjgowers,
         ALescoulie
```

orbeckst commented on package/CHANGELOG:
```diff
@@ -64,6 +68,8 @@ Changes
 
 Deprecations
+  * added `DeprecationWarning` to `analysis.nucleicacids.WatsonCrickDist` when
+    `List[Residue]` is given instead of `ResidueGroup`
```
Write for users:
```suggestion
  * using `List[Residue]` as input for `analysis.nucleicacids.WatsonCrickDist()` is 
    deprecated and will be removed in release 3.0.0; instead use a `ResidueGroup` 
    (Issue #3720, PR #3735)
```

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -183,50 +290,213 @@ class WatsonCrickDist(NucPairDist):
-    ValueError
-        If the residues given are not amino acids
+    DeprecationWarning
+        If a lits of :class:`~MDAnalysis.core.groups.Residue` is given for
```
```suggestion
        If a list of :class:`~MDAnalysis.core.groups.Residue` is given for
```

ALescoulie replied:
done

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -183,50 +290,213 @@ class WatsonCrickDist(NucPairDist):
        MDAnalysis 2.5.0. Please use the class attribute :attr:`times` instead.
+
+    .. versionchanged:: 2.7.0
+        Deprecated the use of :attr:`List[Residue]` instead use
+        :class:`~MDAnalysis.core.groups.ResidueGroup`.
```
This is a bit tricky; normally I would put a `.. deprecated::` role here, but that's going to be too confusing for the casual reader. We're really only deprecating one type of input so I'd rewrite as:
```suggestion
        `strand1` and `strand2` now also accept a :class:`~MDAnalysis.core.groups.ResidueGroup`
        as input. The previous input type of ``List[Residue]`` is still supported
        but it is **deprecated** and will be removed in release 3.0.0.
```

ALescoulie replied:
done

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -183,50 +290,213 @@ class WatsonCrickDist(NucPairDist):
-        If the residues given are not amino acids
+    DeprecationWarning
+        If a lits of :class:`~MDAnalysis.core.groups.Residue` is given for
+        :attr:`strand1` of :attr:`strand2` instead of a
```
Just bare role for arguments
```suggestion
        `strand1` of `strand2` instead of a
```

ALescoulie replied:
I went through and removed the all `attr` used for arguments

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -55,20 +55,34 @@
+from MDAnalysis.core.groups import Residue, ResidueGroup
+
 
+ResidueClass = Union[List[Residue], ResidueGroup]
```
```suggestion
# Deprecation: In 3.0.0 change type to just
# ResidueClass = ResidueGroup
ResidueClass = Union[List[Residue], ResidueGroup]
```

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -55,20 +55,34 @@
+
 
+ResidueClass = Union[List[Residue], ResidueGroup]
+r"""A type alias for :code:`Union[Residue, ResidueGroup]`
```
```suggestion
r"""A type alias for :code:`Union[List[Residue], ResidueGroup]`
```

ALescoulie replied:
done

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -183,50 +290,213 @@ class WatsonCrickDist(NucPairDist):
+        
+        if isinstance(strand1, list) or isinstance(strand2, list):
+            raise DeprecationWarning("ResidueGroup should be used instead of giving a Residue list")
+            strand1: ResidueGroup = ResidueGroup(strand1)
+            strand2: ResidueGroup = ResidueGroup(strand2)
```
This indentation does not look right. Please check.

ALescoulie replied:
The indentation is fine, it appears wrong due to the `DeprecationWarning` going onto the next line. There is an error I overlooked thought, I used `raise` which caused a failure. I switched it to warn so that it just throws a warring then continues running the code. I'll also make sure the code continues working correctly after the warning to be safe.

orbeckst reviewed (changes requested):
There's one potential bug in the code where I noted funny indentation (code never executed after a `raise`).

Please also fix/update the docs for major and minor pair distances.

Finally, please accept my suggestions for rewriting some of the changes.

Almost there! Thanks!

ALescoulie committed:
Update author list in changelog

Co-authored-by: Oliver Beckstein <orbeckst@gmail.com>

ALescoulie committed:
Update features list in changelog

Co-authored-by: Oliver Beckstein <orbeckst@gmail.com>

ALescoulie committed:
Add future decrication notice to WCDist

Co-authored-by: Oliver Beckstein <orbeckst@gmail.com>

Commit:
update documentation with more detailed
descriptions and fix misc typos

Commit:
fix DeprecationWarning in WCDist for case where
strand1 and strand2 are different types

Commit:
add verification helper method to WCDist init to
simplify type checks and catch TypeError

ALescoulie commented:
@orbeckst I believe I addressed your feedback on the documentation, and added some further clarity. I also rewrote how `WatsonCrickDist` handles `List[Resiude]` in order to catch possible errors when both a `List[Residue]` and a `ResidueGroup` are passed in and in cases where the list is not entirely `Residues`. I think I should still add one additional test to get back to 100% coverage as there is one if statement I don't hit, I also will go read the documentation one more time.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -92,21 +108,32 @@ class NucPairDist(AnalysisBase):
 
         .. versionadded:: 2.4.0
 
+    results.distances: numpy.ndarray
+        stored in a 2d numpy array with first index selecting the Residue pair, 
+        and the second index selecting the frame number
```
How is `distances` different from `pair_distances`?

orbeckst replied:
And if it's the same then you need to (sigh) deprecate the first one as it's part of the API.

ALescoulie replied:
I think the change from `pair_distances` to `distances` was a suggestion from above, I added a deprecation notice in the docs and a todo in the code to remove it.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -148,21 +257,31 @@ def _single_frame(self) -> None:
 
     def _conclude(self) -> None:
-        self.results['pair_distances'] = self._res_array
+        self.results['distances'] = self._res_array
```
If you want to change the name then you need to
1. deprecate pair_distances (add entry to CHANGELOG and `.. deprecated` to the entry)
2. keep an entry `self.results['pair_distances'] = self.results['distances']` until it's removed in 3.0.0

The entries in `results` are part of the API so they fall under SemVer.

ALescoulie replied:
OK, I'll add a deprecation notice and add that to the docs and change log

ALescoulie replied:
It's done, I added it to the docs.

orbeckst replied:
Thanks. I rewrote some of the additions using the [.. deprecated:: x.x](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated) role.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -183,50 +302,234 @@ class WatsonCrickDist(NucPairDist):
-    ValueError
-        If the residues given are not amino acids
+    DeprecationWarning
+    .. _Deprecation Warning
```
Does this anchor work? Perhaps a trailing colon missing?

ALescoulie replied:
neither anchor works, I'm having an issues building my docs locally that I asked about on discord. Once I sort that out I'll go over the docs again.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -183,50 +302,234 @@ class WatsonCrickDist(NucPairDist):
        MDAnalysis 2.5.0. Please use the class attribute :attr:`times` instead.
+
+    .. versionchanged:: 2.7.0
+    .. _Deprecation Notice
```
as above : does the anchor work, did you check the generated docs?

orbeckst reviewed (changes requested):
Thanks for addressing my comments. The only real issue is that we need to deprecate `results.pair_distance` in the WatsonCrick class properly.

In the new classes there's nothing to deprecate as they appear here for the first time.

Commit:
Add deprecation notice to docs and changelog for
`pair_distances` in WCDist

Commit:
fix deprecation_warn test and add test for strand
verification TypeError

Commit:
fix docs issues

Commit:
fix formatting

Commit:
simplify tests

ALescoulie commented:
@orbeckst I think this PR about ready, assuming I didn't miss any typos in the docs. I also pulled most of the atom selections in the tests into a fixture so there is less repetition when running the CI.

IAlibay committed:
Merge branch 'develop' into nuclinfo_dist_rebuild

orbeckst committed:
Update CHANGELOG

orbeckst committed:
Update docs in nucleicacids.py

- use deprecated sphinx role
- fixed minor errors in descriptions

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -133,93 +176,428 @@ def __init__(self, selection1: List[mda.AtomGroup],
+    results.distances: numpy.ndarray
+        Distances are stored in a 2d numpy array with axis 0 (first index)
+        indexing the trajectory frame and axis 1 (second index) selecting the  
+        Residue pair.
```
@ALescoulie please check that this rewritten description is correct. It is now consistent with the description of `results.pair_distances`.

orbeckst reviewed (approved):
Thanks @ALescoulie , looks good. 

I made some changes to the docs, mainly to use sphinx markup for deprecations, but I also included a few fixes. In particular, please have a look for yourself that I correctly defined `results.distances`.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -133,93 +176,428 @@ def __init__(self, selection1: List[mda.AtomGroup],
+
+
+class MinorPairDist(NucPairDist):
+    r"""Minor-Pair basepair distance for selected residues
+        over a trajectory.
```
```suggestion
    r"""Minor-Pair basepair distance for selected residues over a trajectory.
```

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -133,93 +176,428 @@ def __init__(self, selection1: List[mda.AtomGroup],
+
+
+class MajorPairDist(NucPairDist):
+    r"""Minor-Pair base pair distance for 
+    selected residues over a trajectory.
```
```suggestion
    r"""Minor-Pair base pair distance for selected residues over a trajectory.
```

orbeckst committed:
Apply suggestions from code review

orbeckst reviewed (changes requested):
In principle all good, but please fix PEP8 things.

ALescoulie commented:
@orbeckst, I'll get it done by tomorrow than I can start on the next part of #3720, the pep8 stuff will just take a few minutes.

Commit:
fix pep8

Commit:
fix pep8 after rebase

Commit:
fix pep8 after rebase

ALescoulie commented:
@orbeckst I got the pep8 stuff done

orbeckst commented:
Hooray. I reopen #3720 for the other things.
